test/distributed/checkpoint/fsdp/test_fsdp_dsd.py

```diff
@@ -182,6 +187,392 @@ def _test_save_with_fsdp1_and_load_with_fsdp2(self, state_dict_type: StateDictTy
+                options=StateDictOptions(full_state_dict=True, cpu_offload=True),
+            )
+
+            # Load state dict into model with FSDP2 + TP applied
+            # FSDP + TP
+            fsdp2_tp_model = _get_base_model()
+            fsdp2_tp_model = parallelize_module(
+                fsdp2_tp_model,
+                device_mesh=tp_mesh,
+                parallelize_plan=tp_parallelize_plan,
+            )
+            for module in fsdp2_tp_model:
+                fully_shard(module, mesh=dp_mesh)
+            fully_shard(fsdp2_tp_model, mesh=dp_mesh)
+            fsdp2_tp_optim = torch.optim.Adam(fsdp2_tp_model.parameters(), lr=0.1)
+
+            fsdp2_tp_state_dict = {
+                "model": get_model_state_dict(fsdp2_tp_model),
+                "optim": get_optimizer_state_dict(fsdp2_tp_model, fsdp2_tp_optim),
+            }
+            dcp.load(
+                fsdp2_tp_state_dict,
+                checkpoint_id=self.temp_dir,
+            )
+            fsdp2_tp_model.load_state_dict(fsdp2_tp_state_dict["model"])
+            fsdp2_tp_optim.load_state_dict(fsdp2_tp_state_dict["optim"])
+
+            fsdp2_tp_full_msd = get_model_state_dict(
+                fsdp2_tp_model,
+                options=StateDictOptions(full_state_dict=True, cpu_offload=True),
+            )
+            fsdp2_tp_full_osd = get_optimizer_state_dict(
+                fsdp2_tp_model,
+                fsdp2_tp_optim,
+                options=StateDictOptions(full_state_dict=True, cpu_offload=True),
+            )
+
+            # Compare full state dict to make sure they are the same.
+            self.assertEqual(base_msd, fsdp2_tp_full_msd)
+            self.assertEqual(base_osd, fsdp2_tp_full_osd)
+            self.assertEqual(fsdp2_tp_full_msd, tp_full_msd)
+            self.assertEqual(fsdp2_tp_full_osd, tp_full_osd)
+
+    @skip_if_lt_x_gpu(4)
+    @with_temp_dir
+    def test_save_with_fsdp2_tp_and_load_with_tp(self):
+        """
+        Test that we can save a model with FSDP2 + TP on 2d mesh and load it with TP.
+        """
+
+        def _get_base_model(mlp_dim: int = 2):
+            base_model = nn.Sequential(MLP(mlp_dim), MLP(mlp_dim), MLP(mlp_dim))
+            return base_model
+
+        tp_parallelize_plan = {
+            "0.in_proj": ColwiseParallel(),
+            "0.out_proj": RowwiseParallel(),
+            "1.in_proj": ColwiseParallel(),
+            "1.out_proj": RowwiseParallel(),
+            "2.in_proj": ColwiseParallel(),
+            "2.out_proj": RowwiseParallel(),
+        }
+
+        # init device mesh
+        dp_size = 2
+        global_mesh_1d = init_device_mesh(
+            "cuda", (self.world_size,), mesh_dim_names=("tp",)
+        )
+        global_mesh_2d = init_device_mesh(
+            "cuda", (dp_size, self.world_size // dp_size), mesh_dim_names=("dp", "tp")
+        )
+        dp_mesh, tp_mesh = global_mesh_2d["dp"], global_mesh_2d["tp"]
+
+        for save_full_state_dict in [True, False]:
```

Comment:
Unlike the previous tests, we may want to preserve both tests because we are testing `full_state_dict` for FSDP2 + TP, which involve strided sharding.